components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -0,0 +1,159 @@
+        })
+    }
+
+    /// This returns a new Locale that is the result of running the
+    /// 'Add Likely Subtags' algorithm from
+    /// https://www.unicode.org/reports/tr35/#Likely_Subtags.
+    pub fn maximize(&self, locale: &mut Locale) -> Result<bool, LocaleCanonicalizerError> {
+        let mut key = LanguageIdentifier {
+            language: locale.language,
+            script: locale.script,
+            region: locale.region,
+            variants: subtags::Variants::default(),
+        };
+
+        let maybe_update_locale = |index: usize, locale: &mut Locale| -> bool {
+            let entry = &self.likely_subtags.entries[index].1;
+            let modified =
+                locale.language.is_empty() || locale.script.is_none() || locale.region.is_none();
+            if locale.language.is_empty() {
+                locale.language = entry.language;
+            }
+            locale.script = locale.script.or(entry.script);
+            locale.region = locale.region.or(entry.region);
+            modified
+        };
+
+        // languages_scripts_regions
+        if locale.script.is_some() && locale.region.is_some() {
+            if let Ok(index) = self
+                .likely_subtags
+                .entries
+                .binary_search_by_key(&&key, |(l, _)| l)
+            {
+                return Ok(maybe_update_locale(index, locale));
+            }
+        }
+
+        // languages_regions
+        if locale.region.is_some() {
+            key.script = None;
+            if let Ok(index) = self
+                .likely_subtags
+                .entries
+                .binary_search_by_key(&&key, |(l, _)| l)
+            {
+                return Ok(maybe_update_locale(index, locale));
+            }
+        }
+
+        // languages_scripts
+        if locale.script.is_some() {
+            key.script = locale.script;
+            key.region = None;
+            if let Ok(index) = self
+                .likely_subtags
+                .entries
+                .binary_search_by_key(&&key, |(l, _)| l)
+            {
+                return Ok(maybe_update_locale(index, locale));
+            }
+        }
+
+        // languages
+        key.script = None;
+        key.region = None;
+        if let Ok(index) = self
+            .likely_subtags
+            .entries
+            .binary_search_by_key(&&key, |(l, _)| l)
+        {
+            return Ok(maybe_update_locale(index, locale));
+        }
+
+        // und_scripts
+        if locale.script.is_some() {
+            key.language = subtags::Language::default();
+            key.script = locale.script;
+            key.region = None;
+            if let Ok(index) = self
+                .likely_subtags
+                .entries
+                .binary_search_by_key(&&key, |(l, _)| l)
+            {
+                return Ok(maybe_update_locale(index, locale));
+            }
+        }
+
+        Err(LocaleCanonicalizerError::NotMatched)
```

Comment:
Why do you treat it as an error? I feel like lack of matching is a perfectly valid outcome from `maximize`. It just means that we don't have anything to extend it to, but it's not an error.

Reply:
The algorithm at https://www.unicode.org/reports/tr35/#Likely_Subtags says to return an error value if there is no match. I like Shane's suggestion of folding everything into an enumeration, so I'll handle things that way.

Reply:
Good point.  UTS 35 says:

> If there is no match,either return
> 1. an error value, or
> 2. the match for "und" (in APIs where a valid language tag is required).

Given the feedback from both Zibi and Mihai, it seems like we should do 2.

Reply:
Ugh, I'm afraid I disagree with both.

I believe in case of failure to match against known likelySubtags we should not alter the locale and return a status that indicates `not matched` (in unic-locale it is `false`, but I'm open to a different one).

The `Result::Err` in rust feels like is designed for a very different paradigm than what LDML is using it for. `maximize` is a request that, by design, may result in no match, since the table of matches is limited. The error would make sense if the operation was mostly called in a case where the user had the right to expect it to succeed, like `parse` should usually parse successfully.
But here we are just asking to maximize if possible and let us know if maximization happened. In my Rust sense, `err` here would make sense if we wanted to return a value out of that scenario, like `binary_search` returns `Err<usize>`. But we don't.
In result I think Error here is not the right call. I think we should go with either bool, or an enum like Shane suggests, although I'm concerned about adding variants to that enum just because we can, and I haven't experienced a case where the information on what fields we modified mattered.
I can imagine some language negotiation optimizing based on that, but for now, I'd be ok if the return enum was:

```
enum LikelySubtagsResult {
    Altered,
    Unaltered
}
```

and if we find use cases for more specific outputs we can extend it.

Reply:
> Ugh, I'm afraid I disagree with both.

Both what?  Both options 1 and 2 from the UTS 35 spec that I mentioned above?

I agree with the idea of returning an enum.  If `.maximize()` can fail for a legitimate reason, then we could return `Result<LikelySubtagsResult, Error>`; otherwise we can just return `LikelySubtagsResult`.

Reply:
> Both what? Both options 1 and 2 from the UTS 35 spec that I mentioned above?

Yeah, both seem like they would lead to bad API design from how I experienced `maximize/minimize` to be used.
I prefer the enum and I'd be in favor of suggesting a change to UTS 35 to not recommend returning an error or matching for `und`.

Reply:
@markusicu @macchiati Can you comment on Zibi's comment above regarding the UTS 35 likely subtags algorithm?

Reply:
We discussed this today in the weekly meeting and the consensus was to go ahead with returning an enum indicating modified/unmodified here.